packages/gatsby/src/datastore/lmdb/query/create-index.ts

```diff
@@ -0,0 +1,309 @@
+import { inspect } from "util"
+import { store } from "../../../redux"
+import { IGatsbyNode } from "../../../redux/types"
+import { IDataStore, ILmdbDatabases } from "../../types"
+import { cartesianProduct, resolveFieldValue } from "./common"
+
+interface IIndexingContext {
+  databases: ILmdbDatabases
+  datastore: IDataStore
+}
+
+export type IndexFields = Map<string, number> // name, direction
+
+export interface IIndexMetadata {
+  state: "ready" | "building" | "stale" | "error" | "initial"
+  error?: string
+  typeName: string
+  keyPrefix: number | string
+  keyFields: Array<[fieldName: string, orderDirection: number]>
+  multiKeyFields: Array<string>
+
+  // Stats for multi-key indexes
+  // (e.g. when node is { id: `id`, foo: [1,2] } it translates into two index keys: [1,`id`], [2,`id`])
+  stats: {
+    keyCount: number
+    itemCount: number
+    maxKeysPerItem: number
+  }
+}
+
+export const undefinedSymbol = Symbol(`undef`)
+
+export type IndexFieldValue =
+  | number
+  | string
+  | boolean
+  | null
+  | typeof undefinedSymbol
+  | Array<IndexFieldValue>
+
+export type IndexKey = Array<IndexFieldValue>
+
+export async function createIndex(
+  context: IIndexingContext,
+  typeName: string,
+  indexFields: IndexFields
+): Promise<IIndexMetadata> {
+  const indexName = buildIndexName(typeName, indexFields)
+  const meta = getIndexMetadata(context, typeName, indexFields, false)
+
+  switch (meta?.state) {
+    case `ready`:
+      return meta
+    case `building`: {
+      return indexReady(context, indexName)
+    }
+    case `initial`:
+    default: {
+      try {
+        await lockIndex(context, indexName)
+      } catch (err) {
+        // Index is being updated in some other process.
+        // Wait and assume it's in a good state when done
+        return indexReady(context, indexName)
+      }
+      return doCreateIndex(context, typeName, indexFields)
+    }
+  }
+}
+
+export function getIndexMetadata(
+  context: IIndexingContext,
+  typeName: string,
+  indexFields: IndexFields,
+  assertReady = true
+): IIndexMetadata {
+  const { databases } = context
+  const indexName = buildIndexName(typeName, indexFields)
+  const meta: IIndexMetadata = databases.metadata.get(toMetadataKey(indexName))
+
+  if (assertReady && meta?.state !== `ready`) {
+    throw new Error(
+      `Index ${indexName} is not ready yet. State: ${meta?.state ?? `unknown`}`
+    )
+  }
+  return meta
+}
+
+async function doCreateIndex(
+  context: IIndexingContext,
+  typeName: string,
+  indexFields: IndexFields
+): Promise<IIndexMetadata> {
+  const { datastore, databases } = context
+  const { indexes, metadata } = databases
+  const indexName = buildIndexName(typeName, indexFields)
+
+  const label = `Indexing ${indexName}`
+  console.time(label)
```

Comment:
Can we use our reporter here instead to show an activity?

Reply:
This is here only while the feature is experimental for debugging purposes. I think we should actually switch to a "single-pass" data analysis before running queries (which will include index creation for all queries). That's where we should add the actual activity.